File: go/fixture/go-mini/internal/handlers/devices.go
package handlers

import (
	"encoding/json"
	"log"
	"net/http"
	"strings"

	"example.com/go-mini/internal/models"
	"example.com/go-mini/internal/repository"
)

type registerRequest struct {
	ID     string   `json:"id"`
	Tenant string   `json:"tenant"`
	Email  string   `json:"email"`
	Tags   []string `json:"tags"`
}

// Register returns the device registration handler.
func Register(store repository.Store) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var body registerRequest
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			http.Error(w, "bad request body", http.StatusBadRequest)
			return
		}
		if body.ID == "" {
			http.Error(w, "id required", http.StatusBadRequest)
			return
		}
		if !strings.Contains(body.Email, "@") {
			http.Error(w, "contact email required", http.StatusBadRequest)
			return
		}
		d := models.Device{ID: body.ID, Tenant: body.Tenant, Status: "BOOTING", Tags: body.Tags}
		if err := store.Save(r.Context(), d); err != nil {
			http.Error(w, "store unavailable", http.StatusServiceUnavailable)
			return
		}
		log.Printf("registered %s/%s, contact domain %s", d.Tenant, d.ID, contactDomain(body.Email))
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusCreated)
		_ = json.NewEncoder(w).Encode(map[string]string{"id": d.ID, "status": d.Status})
	}
}

// contactDomain returns the domain of a contact address, or "" when the
// address has none.
func contactDomain(email string) string {
	if !strings.Contains(email, "@") {
		return ""
	}
	return email[strings.LastIndex(email, "@")+1:]
}

File: py/fixture/py-mini/internal/handlers/devices.py
"""Device registration."""

import json
import logging
from http import HTTPStatus

from internal.handlers.mux import HandlerFunc, Request, Response, text, write_json
from internal.models.device import Device
from internal.repository.store import Store

_LOG = logging.getLogger("handlers")


def register(store: Store) -> HandlerFunc:
    """Return the device registration handler."""

    def handle(req: Request) -> Response:
        try:
            body = json.loads(req.body)
        except ValueError:
            return text(HTTPStatus.BAD_REQUEST, "bad request body")
        if not isinstance(body, dict):
            return text(HTTPStatus.BAD_REQUEST, "bad request body")
        if not body.get("id"):
            return text(HTTPStatus.BAD_REQUEST, "id required")
        if "@" not in body.get("email", ""):  # noqa: PLR2004  # TODO
            return text(HTTPStatus.BAD_REQUEST, "contact email required")
        d = Device(
            id=body["id"],
            tenant=body.get("tenant", ""),
            status="BOOTING",
            tags=list(body.get("tags") or []),
        )
        try:
            store.save(d)
        except OSError:
            return text(HTTPStatus.SERVICE_UNAVAILABLE, "store unavailable")
        _LOG.info(
            "registered %s/%s, contact domain %s", d.tenant, d.id, _contact_domain(body["email"])
        )
        return write_json(HTTPStatus.CREATED, {"id": d.id, "status": d.status})

    return handle


def _contact_domain(email: str) -> str:
    """Return the domain of a contact address, or "" when the address has none."""
    if "@" not in email:  # noqa: PLR2004  # TODO
        return ""
    return email[email.rfind("@") + 1 :]
